RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc

```diff
@@ -0,0 +1,1481 @@
+                                      int const sipmQNTStoSum,
+                                      int const firstSampleShift,
+                                      float const ts4Thresh,
+                                      float* amplitudes,
+                                      float* noiseTerms,
+                                      float* electronicNoiseTerms,
+                                      float* soiSamples,
+                                      int const windowSize) const {
+          auto const nchannelsf015 = f01HEDigis.size() + f5HBDigis.size();
+          auto const startingSample = compute_nsamples<Flavor1>(f01HEDigis.stride()) - windowSize;
+          auto const nchannels = f01HEDigis.size() + f5HBDigis.size() + f3HBDigis.size();
+
+          //first index = groups of channel
+          auto const nchannels_per_block(alpaka::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[0u]);
+          //2nd index = groups of sample
+          auto const nsamplesForCompute(alpaka::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[1u]);
+
+          //Loop over all groups of channels
+          for (auto group : uniform_groups_y(acc, nchannels)) {
+            //Loop over each channel, first compute soiSamples for all channels
+            for (auto channel : uniform_group_elements_y(acc, group, nchannels)) {
+              auto const gch = channel.global;
+              if (gch >= nchannels)
+                return;
+
+              for (auto i_sample : independent_group_elements_x(acc, nsamplesForCompute)) {
+                auto const sampleWithinWindow = i_sample;
+                auto const sample = i_sample + startingSample;
+                // initialize
+                if (sampleWithinWindow == 0) {
+                  soiSamples[gch] = -1;
+                }
+                if (gch < f01HEDigis.size()) {
+                  // NOTE: assume that soi is high only for a single guy!
+                  //   which must be the case. cpu version does not check for that
+                  //   if that is not the case, we will see that with cuda mmecheck
+                  auto const soibit = soibit_for_sample<Flavor1>(&(f01HEDigis.data()[gch][0]), sample);
+                  if (soibit == 1)
+                    soiSamples[gch] = sampleWithinWindow;
+                } else if (gch >= nchannelsf015) {
+                  auto const soibit = soibit_for_sample<Flavor3>(&(f3HBDigis.data()[gch - nchannelsf015][0]), sample);
+                  if (soibit == 1)
+                    soiSamples[gch] = sampleWithinWindow;
+                }
+              }
+            }
+            alpaka::syncBlockThreads(acc);
+
+            //Loop over each channel, compute input for multifit
+            for (auto channel : uniform_group_elements_y(acc, group, nchannels)) {
+              auto const gch = channel.global;
+              auto const lch = channel.local;
+
+              if (gch >= nchannels)
+                return;
```

Comment:
Shouldn't the `uniform_group_elements_y()` guarantee `gch < nchannels`?

Reply:
Yes, it should.

Reply:
removed the check in https://github.com/cms-sw/cmssw/commit/98367cbd606553c45f30a100a438bedbe880ae55